design/validation-spi.md

```diff
@@ -0,0 +1,293 @@
+
+The JSR 380 Bean Validation 2.0 is a mature and rich validation framework that is extensively used in Java applications.
+
+- Optimized for validation of complex Java bean objects based on declarative configuration rather than single values.
+
+Allthough, value validations are possible via `javax.validation.Validator#validateValue`, they require a bean class 
+with a property with the annotated constraints.
+```java
+	<T> Set<ConstraintViolation<T>> Validator#validateValue(Class<T> beanType,
+												  String propertyName,
+												  Object value,
+												  Class<?>... groups);
+```
+[Example for validating a value via Validator#validateValue](https://gist.github.com/thomasdarimont/e6e375819b3b1adff86b7669e5b033b4).
+
+- hibernate-validator is currently not used in Keycloak, thus we'd need to add a dependency for:
+```xml
+<dependency>
+    <groupId>org.hibernate</groupId>
+    <artifactId>hibernate-validator</artifactId>
+    <version>6.0.12.Final</version>
+</dependency>
+```
+- No direct support for validation (Constraint) lookups by name. The user-profile support needs a way to lookup validations by name.
+- No easy way to dynamically configure constraints programatically. The user-profile support needs a way to configure validation dynamically.
+- Definition of [custom validations is quite involved](https://docs.jboss.org/hibernate/stable/validator/reference/en-US/html_single/#validator-customconstraints) - you need 2 classes (annotation + constraint) + error message + META-INF service manifest.
+
+All this was considered and evaluated in a small prototype, but then it was concluded that it makes more sense to go 
+with a custom implementation that provides a narrow API. Additionally custom extensions should be easier and provide 
+a better integration with the existing Keycloak infrastructure. Further more dynamic validation logic lookups 
+and configuration should be possible and easy to do. 
+
+Note that it is still possible to provide a validation/validator implementation that leverages bean validation. 
+Also, bean-validation still works quite well for declarative validation in JAX-RS endpoints.
+
+### Fully fledged Validation SPI
+
+An earlier attempt at a unified validation SPI can be found here [KEYCLOAK-2045 Add support for flexible Validation](https://github.com/keycloak/keycloak/pull/7324) and was discussed on the [keycloak-dev mailing list](https://groups.google.com/g/keycloak-dev/c/-XjQu7rn56s).
+
+After some discussion, it became apparent that this proposal provides a very comprehensive validation framework that 
+offers numerous functionalities that go beyond the requirements of user profile validation and other validations in Keycloak.
+
+However, the API was very complex and not easy to understand. Therefore, it was agreed that @thomasdarimont would revise 
+the proposal again and present it in a smaller form. This has resulted in the new "Simple Validation SPI" proposal.
+
+## Proposed Validation SPI
```

Comment:
I am missing the proposal here. PoC is not sufficient to be a proposal. This document needs to describe what is proposed.

Reply:
Okay, I think this needs some clarification. The Proposed Validation SPI section is more or less an overview and explanation of the [existing Validation SPI PoC](https://github.com/keycloak/keycloak/pull/7887).

I see if I can summarize this a bit here.